dask/layers.py

```diff
@@ -1239,3 +1257,309 @@ def __repr__(self):
         return "DataFrameIOLayer<name='{}', n_parts={}, columns={}>".format(
             self.name, len(self.inputs), self.columns
         )
+
+
+class DataFrameTreeReduction(DataFrameLayer):
+    """DataFrame Tree-Reduction Layer
+
+    Parameters
+    ----------
+    name : str
+        Name to use for the constructed layer.
+    name_input : str
+        Name of the input layer that is being reduced.
+    npartitions_input : str
+        Number of partitions in the input layer.
+    concat_func : callable
+        Function used by each tree node to reduce a list of inputs
+        into a single output value. This function must accept only
+        a list as its first positional argument.
+    tree_node_func : callable
+        Function used on the output of ``concat_func`` in each tree
+        node. This function must accept the output of ``concat_func``
+        as its first positional argument.
+    finalize_func : callable, optional
+        Function used in place of ``tree_node_func`` on the final tree
+        node(s) to produce the final output for each split. By default,
+        ``tree_node_func`` will be used.
+    split_every : int, optional
+        This argument specifies the maximum number of input nodes
+        to be handled by any one task in the tree. Defaults to 32.
+    split_out : int, optional
+        This argument specifies the number of output nodes in the
+        reduction tree. If ``split_out`` is set to an integer >=1, the
+        input tasks must contain data that can be indexed by a ``getitem``
+        operation with a key in the range ``[0, split_out)``.
+    output_partitions : list, optional
+        List of required output partitions. This parameter is used
+        internally by Dask for high-level culling.
+    tree_node_name : str, optional
+        Name to use for intermediate tree-node tasks.
+    """
+
+    name: str
+    name_input: str
+    npartitions_input: str
+    concat_func: Callable
+    tree_node_func: Callable
+    finalize_func: Optional[Callable]
```

Comment:
```suggestion
    concat_func: Callable[[List[T]], CT]
    tree_node_func: Callable[[CT], OT]
    finalize_func: Optional[Callable[[CT], OT]]
```
I _think_ this is accurate? Where `T` probably means `pd.DataFrame` for now. But I could see wanting to generalize this tree reduction layer to arrays, etc. too? Maybe `DataFrameTreeReduction` should be generic on `T, OT` (the input type and output type)?
I personally find these sort of detailed annotations helpful, but maybe others don't.

Reply:
This is another case where the underlying types may be `pd.DataFrame` or `cudf.DataFrame`. Do the `Callable`'s argument types need to be specified to make type-checking useful here? (I don't know much about python type checking)

Reply:
I just liked the idea of noting that `concat_func` produces something that `tree_node_func` and `finalize_func ` have to be able to consume. Even if the input and output types (`T`, `OT`) aren't specified, a type-checker could still validate that part. But it may not be worth it right now.